c/serverStatusService.c

```diff
@@ -0,0 +1,272 @@
+  setDefaultJSONRESTHeaders(response);
+  writeHeader(response);
+  jsonStart(out);
+  if(logLevels){
+    jsonPrintObject(out, logLevels);
+  }
+  jsonEnd(out);
+  finishResponse(response);
+}
+
+void respondWithServerEnvironment(HttpResponse *response, ServerAgentContext *context){
+  /*Information about parameters for smf_unc: https://www.ibm.com/support/knowledgecenter/SSLTBW_2.1.0/com.ibm.zos.v2r1.erbb700/smfp.htm#smfp*/
+  struct utsname unameRet;
+  char hostnameBuffer[256];
+  char* buffer;
+  time_t curr_time;
+  char* c_time_str;
+  int rc = 0;
+  int reqtype = 0x00000005; //fullword. request type
+  int rectype = 0x0000004F; //SMF record type, only type 79 is supported
+  int subtype = 0x00000009; //SMF record subtype
+  int bufferlen = 716800; // Yes the SMF buffer is actually 700Kb roughly
+  int cpuUtil = 0x00000000;
+  int demandPaging = 0x00000000;
+  int options = 0x00000000;
+  int mvsSrm = 0x00000000;
+  int zaapUtil = 0x00000000;
+  int ziipUtil = 0x00000000;
+  uname(&unameRet);
+  gethostname(hostnameBuffer, sizeof(hostnameBuffer));
+  buffer = (char *)safeMalloc(bufferlen, "buffer");
+  memset(buffer, 0, bufferlen);
+  smfFunc = (EXSMFI *)fetch("ERBSMFI");
+  rc = (*smfFunc)(&reqtype,
+               &rectype,
+               &subtype,
+               buffer,
+               &bufferlen,
+               &cpuUtil,
+               &demandPaging,
+               &options,
+               &mvsSrm,
+               &zaapUtil,
+               &ziipUtil);
+  safeFree(buffer, bufferlen);
+  curr_time = time(NULL);
+  c_time_str = ctime(&curr_time);
+  if(c_time_str[strlen(c_time_str) - 1] == '\n'){
+    c_time_str[strlen(c_time_str) - 1] = '\0';
+  }
+  if(rc > 0){
+    respondWithError(response, HTTP_STATUS_BAD_REQUEST, "Unable to fetch from RMF data interface service");
+    return;
+  }
+  jsonPrinter *out = respondWithJsonPrinter(response);
+  setResponseStatus(response, 200, "OK");
+  setDefaultJSONRESTHeaders(response);
+  writeHeader(response);
+  jsonStart(out);
+  jsonAddString(out, "timestamp", c_time_str);
+  if(getenv("ZSS_LOG_FILE") != NULL){
+    jsonAddString(out, "logDirectory", getenv("ZSS_LOG_FILE"));
+  } else {
+    jsonAddString(out, "logDirectory", "ZSS_LOG_FILE not defined in environment variables");
+  }
+  jsonAddString(out, "agentName", "zss");
+  jsonAddString(out, "agentVersion", context->productVersion);
+  jsonAddString(out, "arch", unameRet.sysname);
+  jsonAddString(out, "osRelease", unameRet.release);
+  jsonAddString(out, "osVersion", unameRet.version);
+  jsonAddString(out, "hardwareIdentifier", unameRet.machine);
+  jsonAddString(out, "hostname", hostnameBuffer);
+  jsonAddString(out, "nodename", unameRet.nodename);
+  jsonStartObject(out, "userEnvironment");
+  char *env_var = *environ;
+  for (int i = 1; env_var; i++) {
+    char *equalSign = strchr(env_var, '=');
+    if(equalSign == NULL){
+      break;
+    }
+    int nameLen = strlen(env_var) - strlen(equalSign);
```

Comment:
If you could find the maximum length of this on z/OS, then you could allocate it on the stack instead.